File: 146.lru-缓存.py
#
# @lc app=leetcode.cn id=146 lang=python3
#
# [146] LRU 缓存
#

# @lc code=start
class ListNode:
    
    def __init__(self,_key=0,_value=0):
        self.key=_key
        self.value=_value
        self.pre=None 
        self.next=None
class LRUCache:

        
    def __init__(self, capacity: int):
        self.capacity=capacity
        self.cache={}
        self.head=ListNode()
        self.tail=ListNode()
        self.head.next=self.tail 
        self.tail.pre=self.head
    def add_head(self,node):
        node.pre=self.head
        node.next=self.head.next
        self.head.next.pre=node
        self.head.next=node 
    def remove_node(self,node):
        node.pre.next=node.next
        node.next.pre=node.pre
    def remove_tail(self):
        tail = self.tail.pre 
        self.remove_node(tail)
        return tail 
    def move_head(self,node):
        self.remove_node(node)
        self.add_head(node)

        

        

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.move_head(node)
            return node.value
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.value=value 
            self.move_head(node)
        else:
            node = ListNode(key,value)
            self.cache[key]=node 
            self.add_head(node)
            if len(self.cache)>self.capacity:
                k = self.remove_tail().key
                self.cache.pop(k)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

File: 146.lru-缓存.java
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=146 lang=java
 *
 * [146] LRU 缓存
 */

// @lc code=start
class LRUCache {

    class DeLinkNode{
        int key,value;
        DeLinkNode pre,next;
        public DeLinkNode(int _key,int _value){
            key=_key;
            value=_value;
        }
        public DeLinkNode(){};
    }

    DeLinkNode head,tail;
    int size,capacity;
    HashMap<Integer,DeLinkNode> cache = new HashMap<>();

    public LRUCache(int _capacity) {
        size=0;
        capacity=_capacity;
        head = new DeLinkNode();
        tail = new DeLinkNode();
        head.next = tail;
        tail.pre = head;
    }
    
    public int get(int key) {
        if(!cache.containsKey(key))
            return -1;
        DeLinkNode node = cache.get(key);
        moveFirst(node);
        return node.value;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)){
            DeLinkNode node = cache.get(key);
            moveFirst(node);
            node.value=value;
        }
        else{
            DeLinkNode node = new DeLinkNode(key,value);
            cache.put(key, node);
            addFirst(node);
            size++;
            if(size>capacity){
                int keyTail = removeTail().key;
                cache.remove(keyTail);
                size--;
            }
        }
    }

    public void removeNode(DeLinkNode node){
        node.pre.next = node.next;
        node.next.pre = node.pre;
    }
    public void addFirst(DeLinkNode node){
        node.pre=head;
        node.next=head.next;
        head.next.pre=node;
        head.next=node;
    }
    public void moveFirst(DeLinkNode node){
        removeNode(node);
        addFirst(node);
    }
    public DeLinkNode removeTail(){
        DeLinkNode last = tail.pre;
        removeNode(last);
        return last;
    }

    // hashMap + deLink
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
// @lc code=end

